File: FCLCore/src/main/java/com/tungsten/fclcore/mod/modinfo/QuiltModMetadata.java
package com.tungsten.fclcore.mod.modinfo;

import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.tungsten.fclcore.mod.LocalModFile;
import com.tungsten.fclcore.mod.ModLoaderType;
import com.tungsten.fclcore.mod.ModManager;
import com.tungsten.fclcore.util.gson.JsonUtils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.stream.Collectors;

public final class QuiltModMetadata {
    private record QuiltLoader(String id, String version, Metadata metadata) {
        private record Metadata(String name, String description, JsonObject contributors,
                                String icon, JsonObject contact) {
        }

    }

    private final int schema_version;
    private final QuiltLoader quilt_loader;

    public QuiltModMetadata(int schemaVersion, QuiltLoader quiltLoader) {
        this.schema_version = schemaVersion;
        this.quilt_loader = quiltLoader;
    }

    public static LocalModFile fromFile(ModManager modManager, Path modFile, FileSystem fs) throws IOException, JsonParseException {
        Path path = fs.getPath("quilt.mod.json");
        if (Files.notExists(path)) {
            throw new IOException("File " + modFile + " is not a Quilt mod.");
        }

        QuiltModMetadata root;
        try (InputStream is = Files.newInputStream(path)) {
            root = JsonUtils.fromNonNullJsonFully(is, QuiltModMetadata.class);
        }
        if (root.schema_version != 1) {
            throw new IOException("File " + modFile + " is not a supported Quilt mod.");
        }

        return new LocalModFile(
                modManager,
                modManager.getLocalMod(root.quilt_loader.id, ModLoaderType.QUILT),
                modFile,
                root.quilt_loader.metadata.name,
                new LocalModFile.Description(root.quilt_loader.metadata.description),
                root.quilt_loader.metadata.contributors.entrySet().stream().map(entry -> String.format("%s (%s)", entry.getKey(), entry.getValue().getAsJsonPrimitive().getAsString())).collect(Collectors.joining(", ")),
                root.quilt_loader.version,
                "",
                Optional.ofNullable(root.quilt_loader.metadata.contact.get("homepage")).map(jsonElement -> jsonElement.getAsJsonPrimitive().getAsString()).orElse(""),
                root.quilt_loader.metadata.icon
        );
    }
}

File: FCL/src/main/java/com/tungsten/fclcore/mod/modinfo/QuiltModMetadata.kt
@file:OptIn(kotlinx.serialization.InternalSerializationApi::class)

package com.tungsten.fclcore.mod.modinfo

import com.tungsten.fclcore.mod.LocalModFile
import com.tungsten.fclcore.mod.ModLoaderType
import com.tungsten.fclcore.mod.ModManager
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonPrimitive
import com.tungsten.fclcore.util.tree.ZipFileTree
import java.io.IOException
import java.nio.file.Path

@Serializable
data class QuiltModMetadata(
    @SerialName("schema_version")
    val schemaVersion: Int = 0,
    @SerialName("quilt_loader")
    val quiltLoader: QuiltLoader
) {
    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun fromFile(modManager: ModManager, modFile: Path, tree: ZipFileTree): LocalModFile {
            val path = tree.getEntry("quilt.mod.json")
                ?: throw IOException("File $modFile is not a Quilt mod.")

            val root: QuiltModMetadata = MOD_METADATA_JSON.decodeFromString(tree.readTextEntry(path))
            if (root.schemaVersion != 1)
                throw IOException("File $modFile is not a supported Quilt mod.")

            val loader = root.quiltLoader
            val metadata = loader.metadata
            val contributors = metadata.contributors.entries.joinToString(", ") { (name, value) ->
                "$name (${value.jsonPrimitive.content})"
            }
            return LocalModFile(
                modManager,
                modManager.getLocalMod(loader.id, ModLoaderType.QUILT),
                modFile,
                metadata.name,
                LocalModFile.Description(metadata.description),
                contributors,
                loader.version,
                "",
                (metadata.contact["homepage"] as? JsonPrimitive)?.content ?: "",
                metadata.icon
            )
        }
    }
}

@Serializable
data class QuiltLoader(
    val id: String = "",
    val version: String = "",
    val metadata: QuiltMetadata = QuiltMetadata()
)

@Serializable
data class QuiltMetadata(
    val name: String = "",
    val description: String = "",
    val contributors: Map<String, JsonElement> = emptyMap(),
    val icon: String = "",
    val contact: Map<String, JsonElement> = emptyMap()
)
